subsys/zigbee/cli/zigbee_cli_cmd_zcl_groups.c

```diff
@@ -0,0 +1,996 @@
+		    (packet_info->dst_ep == ZB_APS_BROADCAST_ENDPOINT_NUMBER)) {
+			return ZB_FALSE;
+		} else {
+			return ZB_TRUE;
+		}
+	default:
+		return ZB_FALSE;
+	}
+}
+
+/**@brief Add remote node to the group or remove from the group.
+ *
+ * @code
+ * zcl group add <h:dst_addr> <d:ep> [-p h:profile] <h:group_id>
+ * @endcode
+ *
+ * Send Add Group command to the endpoint `ep` of the remote node `dst_addr`
+ * and add it to the group `group_id`.
+ * Providing profile ID is optional, Home Automation Profile is used by default.
+ *
+ * @code
+ * zcl group remove <h:dst_addr> <d:ep> [-p h:profile] <h:group_id>
+ * @endcode
+ *
+ * Send Remove Group command to the endpoint `ep` of the remote node `dst_addr`
+ * and remove it from the group `group_id`.
+ * Providing profile ID is optional, Home Automation Profile is used by default.
+ *
+ * @note These commands can be sent as groupcast. To send command as groupcast,
+ *       set `dst_addr` to group address and `ep` to 0.
+ *
+ * @note Group name is not supported.
+ */
+int cmd_zb_add_remove_group(const struct shell *shell, size_t argc, char **argv)
+{
+	zb_bufid_t bufid;
+	struct groups_cmd_data *req_data;
+	struct zcl_packet_info *packet_info;
+	struct ctx_entry *entry = ctx_mgr_new_entry(CTX_MGR_GROUPS_CMD_ENTRY_TYPE);
+
+	if (entry == NULL) {
+		zb_cli_print_error(shell, "Request pool empty - wait a bit", ZB_FALSE);
+		return -ENOEXEC;
+	}
+
+	req_data = &(entry->groups_cmd_data);
+	packet_info = &(req_data->packet_info);
+
+	/*  Set default profile ID to Home Automation. */
+	packet_info->prof_id = ZB_AF_HA_PROFILE_ID;
+	/* Set function to be used to send Add Group or Remove Group command. */
+	if (strcmp(*(argv), "add") == 0) {
+		req_data->send_fn = zb_zcl_send_add_group_cmd;
+	} else if (strcmp(*(argv), "remove") == 0) {
+		req_data->send_fn = zb_zcl_send_remove_group_cmd;
+	} else {
+		zb_cli_print_error(entry->shell, "Failed to parse command name", ZB_FALSE);
+		goto add_remove_group_error;
+	}
+
+	packet_info->dst_addr_mode = parse_address(*(++argv), &(packet_info->dst_addr), ADDR_ANY);
+
+	if (packet_info->dst_addr_mode == ADDR_INVALID) {
+		zb_cli_print_error(shell, "Invalid address", ZB_FALSE);
+		goto add_remove_group_error;
+	}
+
+	if (!zb_cli_sscan_uint8(*(++argv), &(packet_info->dst_ep))) {
+		zb_cli_print_error(entry->shell, "Incorrect remote endpoint", ZB_FALSE);
+		goto add_remove_group_error;
+	}
+
+	/* Handle sending to group address. */
+	if ((packet_info->dst_addr_mode == ZB_APS_ADDR_MODE_16_ENDP_PRESENT)
+	    && (packet_info->dst_ep == 0)) {
+		zb_uint16_t group_addr = packet_info->dst_addr.addr_short;
+
+		/* Verify group address. */
+		if ((group_addr < ZB_ZCL_ATTR_SCENES_CURRENT_GROUP_MIN_VALUE) ||
+		    (group_addr > ZB_ZCL_ATTR_SCENES_CURRENT_GROUP_MAX_VALUE)) {
+			zb_cli_print_error(shell, "Incorrect group address", ZB_FALSE);
+			goto add_remove_group_error;
+		}
+
+		packet_info->dst_addr_mode = ZB_APS_ADDR_MODE_16_GROUP_ENDP_NOT_PRESENT;
```

Comment:
What do you think about adding this logic into the `parse_address` utility -or- creating a new, extended variant of it?

Reply:
As soon I can get into refactoring Zigbee shell I would like to do that ;)